### PR TITLE
feat: introduce the turborepo to the parser.js

### PR DESCRIPTION
**Pull request by ayushnau**
Changes Added.

    Turborepo Added.
    Converted the parser repo to app, inside the turborepo.
    Added the required scripts of the turborepo to the root package.json file.

Related issue(s)

Related to: https://github.com/asyncapi/parser-js/issues/963

### CODE REVIEW

**Comment by ayushnau** [comment]
Deliverables:
1. Integrate turborepo in parser.js
2. Integrate multiparser in the parser.js.

**Comment by ayushnau** [comment]
@smoya please review the pr.

**Comment by smoya** [comment]
Blocking the merge until we review it

File: README.md
Comment (smoya) [inline README.md:1]: We need a proper README. This is just a placeholder.

File: apps/parser/CONTRIBUTING.md
Comment (smoya) [inline apps/parser/CONTRIBUTING.md:1]: This file should go in the root of the repository

File: apps/parser/LICENSE
Comment (smoya) [inline apps/parser/LICENSE:1]: This file should go in the root of the repository

File: apps/parser/.gitignore
Comment (smoya) [inline apps/parser/.gitignore:32]: Why adding so much things to the .gitignore? What are those additions based on?

For example, the `.next` dir?

File: turbo.json
Comment (smoya) [inline turbo.json:6]: Why next?

File: apps/parser/CODEOWNERS
Comment (AyushNautiyalDeveloper) [reply apps/parser/CODEOWNERS:1]: added in root.

File: apps/parser/CODE_OF_CONDUCT.md
Comment (AyushNautiyalDeveloper) [reply apps/parser/CODE_OF_CONDUCT.md:1]: added in root.

File: apps/parser/CONTRIBUTING.md
Comment (AyushNautiyalDeveloper) [reply apps/parser/CONTRIBUTING.md:1]: added in root.

File: apps/parser/.gitignore
Comment (AyushNautiyalDeveloper) [reply apps/parser/.gitignore:32]: removed unwanted keys.

File: turbo.json
Comment (AyushNautiyalDeveloper) [reply turbo.json:6]: removed.

File: README.md
Comment (smoya) [inline README.md:1]: In one side, please revert this file back to the version in `master` branch. You are not making additions but just styling changes, which are not related to the work to do in this PR.

On the other side, you should have one README in the root, mentioning this repository is a monorepo that contains several apps (only parser-js atm), similar to https://github.com/asyncapi/studio. And as nice to have, I would say a link to the README's of each app.

File: apps/parser/LICENSE
Comment (smoya) [inline apps/parser/LICENSE:1]: Move this file to the root of the repo.

File: turbo.json
Comment (smoya) [reply turbo.json]: Did you test the build?

**Comment by smoya** [comment]
Please change the prefix of the PR to `refactor:` or `chore:`. Otherwise, we would be releasing a new version of the parser-js as if it was a feature when it's not.

Additionally, edit the `Resolves #963` part in the issue description to something like `Related to #963` since this PR is not solving it fully. Otherwise, the issue will be closed automatically once this PR gets merged.

**Comment by smoya** [comment]
What do you think are the next steps? Shall we manually test this? Like trying to run a dry-run release.

**Comment by smoya** [comment]
@aayushmau5 please rename the PR accordingly to https://github.com/asyncapi/parser-js/pull/992#issuecomment-2085224668

**Comment by smoya** [comment]
@ayushnau How are we gonna build the parser from now on? Should `npm run build` just work? At this moment, it gives you the following error (it's true, no `build` task).

```bash
❯ npm run build

> build
> turbo run build

  × missing tasks in project

Error:   × could not find task `build` in project
```

Shall we additionally change any path (if it changes) or command in the GH actions workflows? Have you tried to run semantic-release in dry-run mode?

**Comment by ayushnau** [comment]
> @ayushnau How are we gonna build the parser from now on? Should `npm run build` just work? At this moment, it gives you the following error (it's true, no `build` task).
> 
> ```shell
> ❯ npm run build
> 
> > build
> > turbo run build
> 
>   × missing tasks in project
> 
> Error:   × could not find task `build` in project
> ```
> 
> Shall we additionally change any path (if it changes) or command in the GH actions workflows? Have you tried to run semantic-release in dry-run mode?

till other repo is not merged "npm run build" is fine after that will add the command for parser only build "npm run parser:build" like this.

**Comment by smoya** [comment]
For the record, as I shared with you @ayushnau via Slack DM, the `npm run build` command is failing on my computer. I'm using node v18. There is something I'm doing wrong? 

image

**Comment by smoya** [comment]
@ayushnau Did you have a chance to test it? 🙏

**Comment by ayushnau** [comment]
I looked into it, 

> @ayushnau Did you have a chance to test it? 🙏

@smoya which node version are you using.  Cant find the exact reason but most probably it is happening because of node version. Mine is version 20.

**Comment by smoya** [comment]
> I looked into it,
> 
> > @ayushnau Did you have a chance to test it? 🙏
> 
> @smoya which node version are you using. Cant find the exact reason but most probably it is happening because of node version. Mine is version 20.

No, it wasn't the version of Node or NPM, but rather I had to manually remove both `node_modules` dir from root and `apps/parser` dirs.

**Comment by smoya** [comment]
@ayushnau I believe next step should be to emulate a release by running `npx semantic-release --dry-run`. You will require to set tokens, I recommend you create your owns. In case of pushing to npm, you could change the package scope so you can push to your own npm account.

**Comment by ayushnau** [comment]
image
image

@smoya The release is working fine in dry run.

**Comment by aeworxet** [comment]
@asyncapi/bounty_team

**Comment by smoya** [comment]
> image
> image
> 
> @smoya The release is working fine in dry run. 

As you are aware @ayushnau, im off this week, so I cant merge this PR until im back.
This PR contains critical changes so it requires a controlled and watched merge and release.

**Comment by smoya** [comment]
@jonaslagoni @magicmatatjahu would you have a chance to do a review on this one before next week?🙏

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
We have a huge regression of the documentation, root readme files etc, that should not happen with the rewrite unless it's to a temporary branch 🙂

**Comment by ayushnau** [comment]
@smoya @jonaslagoni I have updated the readmd file. 
Also other than that I have added the multiparser in the turborepo. working on pointing the multiparser to local parser repo.

File: README.md
Comment (smoya) [inline README.md:3]: Please, merge all of this new docs (or remove if repeated) into the previous README contents.

As you can see, and as @jonaslagoni mentioned, we don't want to have a worse documentation than what we have. It's about the format, the header of the repository, the TOC, etc. 
Google Chrome_jjqAjmOb

**Comment by smoya** [comment]
> Also other than that I have added the multiparser in the turborepo. (ayushnau:feat-multiparser branch) working on pointing the multiparser to local parser repo.

If you want us to review, please open a PR.

**Review by smoya — CHANGES_REQUESTED** [review]
https://github.com/asyncapi/parser-js/pull/992#discussion_r1615754713

**Comment by smoya** [comment]
> I encourage you to do a proper release without the dry-run. In that case, before pushing to npm, you could change the package scope so you can push to your own npm account.

@ayushnau have you ever tested that?

**Comment by AyushNautiyalDeveloper** [comment]
> > I encourage you to do a proper release without the dry-run. In that case, before pushing to npm, you could change the package scope so you can push to your own npm account.
> 
> @ayushnau have you ever tested that?

No I have done the test for only dry run.

**Comment by smoya** [comment]
> No I have done the test for only dry run.

I feel it is totally required to do this before merging.

**Comment by ayushnau** [comment]
@smoya updated the readme.md file. 
also here is the pr for multiparser for review https://github.com/ayushnau/parser-js/pull/1 check.

File: README.md
Comment (smoya) [inline README.md:26]: The TOC content is autogenerated, so please revert them back to previous values.

File: README.md
Comment (smoya) [inline README.md:12]:
```suggestion
1. [Parser-js](apps/parser): Use this package to validate and parse AsyncAPI documents —either YAML or JSON— in your Node.js or browser application.
```

File: README.md
Comment (smoya) [inline README.md:10]: I believe this should be moved up and somehow replace content from line 3 and beyond (but keeping all the details about Spectral, etc).

**Comment by ayushnau** [comment]
image
image

@smoya it is working fine with 
`npx semantic-release` it is working fine but only if I am making a small change. 

image
here there in node_modules dependency of `conventional-changelog-conventionalcommits` that is basically checking for the dates. but the dates it is recieving for all the commits is `{}` which is not getting return. hence causing error. 

So now My question making changes in the node_modules is possible here?

**Comment by smoya** [comment]
> This dependency `conventional-changelog-conventionalcommits` is installed by me to run the for running semantic release. will be need this dependency to be present there or not.

@ayushnau I'm sorry but after reading several times I didn't really get if this is a question regarding the need of such package (we need it for conventional commits) or that you are saying the culprit is such a package. What is exactly the "small change" you refer to?

**Comment by AyushNautiyalDeveloper** [comment]
> > This dependency `conventional-changelog-conventionalcommits` is installed by me to run the for running semantic release. will be need this dependency to be present there or not.
> 
> @ayushnau I'm sorry but after reading several times I didn't really get if this is a question regarding the need of such package (we need it for conventional commits) or that you are saying the culprit is such a package. What is exactly the "small change" you refer to?

1. for testing `npx semantic-realease` I was needed to install `conventional-changelog-conventionalcommits` package. Now is it alright or not. As this package is not the part of the parser-js. I used it just for testing purposes.
2. if yes, while using this I was required to make some changes in the node_modules of this package. Now can you give me the comment on this. Overall if this is correct or not.

**Comment by ayushnau** [comment]
> > This dependency `conventional-changelog-conventionalcommits` is installed by me to run the for running semantic release. will be need this dependency to be present there or not.
> 
> @ayushnau I'm sorry but after reading several times I didn't really get if this is a question regarding the need of such package (we need it for conventional commits) or that you are saying the culprit is such a package. What is exactly the "small change" you refer to?

1. for testing `npx semantic-realease` I was needed to install `conventional-changelog-conventionalcommits` package. Now is it alright or not. As this package is not the part of the parser-js. I used it just for testing purposes.
2. if yes, while using this I was required to make some changes in the node_modules of this package. Now can you give me the comment on this. Overall if this is correct or not.

**Comment by smoya** [comment]
> 1. for testing `npx semantic-realease` I was needed to install `conventional-changelog-conventionalcommits` package. Now is it alright or not. As this package is not the part of the parser-js. I used it just for testing purposes.

That package is installed by the GH action that runs the release. See https://github.com/asyncapi/parser-js/blob/05b782b1ae4019a5ecf41ab0a2984a660b40114c/.github/workflows/if-nodejs-release.yml#L110

So you don't need to install it.

> 2. if yes, while using this I was required to make some changes in the node_modules of this package. Now can you give me the comment on this. Overall if this is correct or not.

I don't expect any change is needed to be done. What are the changes you needed to do?

**Comment by ayushnau** [comment]
add Object.keys condition 
image

**Comment by smoya** [comment]
> added the  Object.keys condition in the node_modules of the package. Otherwise I am getting error. @smoya 
> 
> image
> 

But you mean in the `conventional-changelog-conventionalcommits` package?
We do not.own such a package. Are you sure we are doing things right? Maybe some misconfiguration?

**Comment by ayushnau** [comment]
@smoya It is working fine now. I changed the version and now it is working. 
image

**Review by smoya — DISMISSED** [review]
LGTM! 🚀🌔

**Comment by smoya** [comment]
@ayushnau There is an error, seems a not updated .lock file or similar https://github.com/asyncapi/parser-js/actions/runs/9291475481/job/25618079508?pr=992

**Comment by ayushnau** [comment]
@smoya updated can you check now.

**Comment by smoya** [comment]
There is [an issue with our tests](https://github.com/asyncapi/parser-js/actions/runs/9348904713?pr=999) in CI when running under macos. It is not the first time we find this. 
We need to investigate what the issue is, but not linked to this PR.

**Comment by AyushNautiyalDeveloper** [comment]
ok sure @smoya

**Comment by ayushnau** [comment]
@smoya please check i have updated the pr.

**Comment by smoya** [comment]
@jonaslagoni can you give your 👍 now?

**Review by jonaslagoni — APPROVED** [review]
@smoya if you are happy with it, go for it 👍

**Comment by smoya** [comment]
> @smoya if you are happy with it, go for it 👍 Dont see anything from my side

I only have one concern. I want to merge this ASAP and release it (so we can be sure everything works as expected, including CI pipeline etc). The point is that this is a `chore` and atm I don't see any opened PR with a mergeable status we could merge right after this one in order to do a meaningful release. 
Forcing a release for this without providing any new feature or fix is not very desirable... 

Any idea or suggestion @jonaslagoni ?

**Comment by jonaslagoni** [comment]
I think it's fine regardless, do what is easiest don't think too much about it 🙂

**Comment by smoya** [comment]
I'm gonna merge this as a new minor version (`feat`). I prefer to fail fast than cry late.

**Comment by smoya** [comment]
> I'm gonna merge this as a new minor version (`feat`). I prefer to fail fast than cry late.

Well, failed fast 😆 

> [7:50:15 AM] [semantic-release] [@semantic-release/npm] › ℹ  Skip publishing to npm registry as package.json's private property is true

See https://github.com/asyncapi/parser-js/actions/runs/9413515248/job/25930521532#step:9:65

@ayushnau I believe you tested this as per your previous comments (to publish to your own NPM account). Anyway, we need to fix this asap as `3.1.0` is published only in GH but not in NPM :\

**Comment by smoya** [comment]
The issue comes with semantic-release and it's lack of support for monorepo. We should move into [Changesets](https://github.com/changesets/changesets) as Studio did: https://github.com/asyncapi/studio/blob/master/doc/adr/0002-use-changesets-for-publishing-and-versioning.md

I'm reverting this PR. @ayushnau Feel free to open a new one with a version that really releases each package individually.

**Pull request merged**
